lib/class-wp-rest-batch-controller.php

```diff
@@ -0,0 +1,298 @@
+<?php
+/**
+ * REST API: WP_REST_Batch_Controller class
+ *
+ * @package    Gutenberg
+ * @subpackage REST_API
+ */
+
+/**
+ * Core class used to perform abtch requests.
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Batch_Controller {
+
+	/**
+	 * Registers the REST API route.
+	 *
+	 * @since 9.0.0
+	 */
+	public function register_routes() {
+		register_rest_route(
+			'__experimental',
+			'batch',
+			array(
+				'callback'            => array( $this, 'serve_batch_request' ),
+				'permission_callback' => '__return_true',
+				'methods'             => array( 'POST', 'PUT', 'PATCH', 'DELETE' ),
+				'args'                => array(
+					'validation' => array(
+						'type'    => 'string',
+						'enum'    => array( 'require-all-validate', 'normal' ),
+						'default' => 'normal',
+					),
+					'requests'   => array(
+						'required' => true,
+						'type'     => 'array',
+						'maxItems' => 25,
+						'items'    => array(
+							'type'       => 'object',
+							'properties' => array(
+								'path'    => array(
+									'type'     => 'string',
+									'required' => true,
+								),
+								'body'    => array(
+									'type'                 => 'object',
+									'properties'           => array(),
+									'additionalProperties' => true,
+								),
+								'headers' => array(
+									'type'                 => 'object',
+									'properties'           => array(),
+									'additionalProperties' => array(
+										'type'  => array( 'string', 'array' ),
+										'items' => array(
+											'type' => 'string',
+										),
+									),
+								),
+							),
+						),
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Serves the batch request.
+	 *
+	 * @since 9.0.0
+	 *
+	 * @param WP_REST_Request $batch_request The batch request object.
+	 * @return WP_REST_Response
+	 */
+	public function serve_batch_request( WP_REST_Request $batch_request ) {
+		$requests = array();
+
+		foreach ( $batch_request['requests'] as $args ) {
+			$parsed_url = wp_parse_url( $args['path'] );
+
+			if ( false === $parsed_url ) {
+				$requests[] = new WP_Error( 'parse_path_failed', __( 'Could not parse the path.', 'gutenberg' ), array( 'status' => 400 ) );
+
+				continue;
+			}
+
+			$single_request = new WP_REST_Request( $batch_request->get_method(), $parsed_url['path'] );
+
+			if ( ! empty( $parsed_url['query'] ) ) {
+				$query_args = null; // Satisfy linter.
+				wp_parse_str( $parsed_url['query'], $query_args );
+				$single_request->set_query_params( $query_args );
+			}
+
+			if ( ! empty( $args['body'] ) ) {
+				$single_request->set_body_params( $args['body'] );
+			}
+
+			$requests[] = $single_request;
```

Comment:
Schema allows `headers` to be a part of each request so we should hydrate each `$single_request` with user-provided headers.

Reply:
🤦 Good catch. Added in 7471d6ddee.